tests/periph_gpio: added benchmark capabilities

haukepetersen opened the pull request:
alternative to #7478

Instead of adding a dedicated benchmark application this PR simply adds the same functionality as shell command to the existing `periph_gpio` test application (as done some time ago also for the `periph_spi` test). For convenience and potentially more advanced improvements I added also a `make bench` target using pexpect. This way one can (semi-) automatically run the benchmark on any platform.

I know there is a 1000 things that can be build upon this (and possible also improved), but I'd like to get this in in the current state for now and we can always improve this later. But it is quite useful already and I prefer this solution over #7478.

kaspar030 commented on tests/periph_gpio/main.c:
```diff
@@ -203,6 +207,97 @@ static int toggle(int argc, char **argv)
+    gpio_t pin = GPIO_PIN(atoi(argv[1]), atoi(argv[2]));
+    unsigned long runs = BENCH_RUNS_DEFAULT;
+    if (argc > 3) {
+        runs = (unsigned long)atoi(argv[3]);
```
The return value needs to be checked (for >0). Otherwise, later on, this can lead to div by zero. (atio() returns 0 for non-number strings...)

kaspar030 replied:
And maybe use atol() or atoll() so 16bitter can do more than 2^15?

kaspar030 commented on tests/periph_gpio/main.c:
```diff
@@ -203,6 +207,97 @@ static int toggle(int argc, char **argv)
+    unsigned state = irq_disable();
+    time = xtimer_now_usec();
+    for (unsigned long i = 0; i < runs; i++) {
+        __asm("nop");
```
```__asm__```, and maybe add volatile?

kaspar030 commented on tests/periph_gpio/main.c:
```diff
@@ -203,6 +207,97 @@ static int toggle(int argc, char **argv)
+    irq_restore(state);
+    bench_print_time(time, runs, "gpio_read");
+
+
```
remove extra spaces

kaspar030 commented on tests/periph_gpio/main.c:
```diff
@@ -203,6 +207,97 @@ static int toggle(int argc, char **argv)
+    state = irq_disable();
+    time = xtimer_now_usec();
+    for (unsigned long i = 0; i < runs; i++) {
+        int tmp = gpio_read(pin);
```
why save the return and then void it? does it prevent optimizing the function call away?

kaspar030 commented on tests/periph_gpio/main.c:
```diff
@@ -203,6 +207,97 @@ static int toggle(int argc, char **argv)
+    bench_print_time(time, runs, "gpio_write");
+
+    /* gpio_read */
+    state = irq_disable();
```
would it, in this case, make sense to create a ```BENCHMARK(name, runs, command)``` macro?

haukepetersen commented:
Addressed comments and introduced an independent benchmark module for running this type of trivial function timing benchmarks. This makes the code much nicer to read, indeed.

Making use of the benchmak modules in other tests (e.g. `periph_spi`) is open for follow up PRs...

gebart commented:
I have been thinking about something similar to this benchmark module using the Cortex-M systick timer to measure actual cycles. However, that would be a platform specific hack if I ever get around to implementing it. It could be useful to be able to measure the overhead in ISRs or other low level functionality.

haukepetersen commented:
I think for the Cortex tick counter, it would make sense to write a (sub) module to the cortexm_common code, which allows to use this capability in an easy way. This would indeed be very useful thing!

haukepetersen commented:
rebased.

@kaspar030 what do you say?

kaspar030 commented on sys/benchmark/benchmark.c:
```diff
@@ -0,0 +1,31 @@
+    uint32_t div  = (time - (full * runs)) / (runs / 1000);
+
+    printf("%11s: %9luus  ---  %2u.%03uus per call\n", name,
+           (unsigned long) time, (unsigned)full, (unsigned)div);
```
better use ```PRIu32``` et al ...

kaspar030 commented on sys/benchmark/benchmark.c:
```diff
@@ -0,0 +1,31 @@
+ * @}
+ */
+
+#include <stdio.h>
```
here and below: usually we put a new line after system headers

haukepetersen replied:
I usually do this, too, but in cases where there is only a small (roughly up to 2 or 3 ) includes in a file I tend to put them in a single block, as I think its quite ugly to have a single include per 'newline separated block'... But will fix this here to make you happy :-)

kaspar030 commented:
Tested, works flawlessly. Feel free to directly squash the printf fixes.

haukepetersen commented:
rebased, squashed, and fixed style and printf issues pointed out above.

haukepetersen commented:
triggered Mudock (forgot earlier...)

haukepetersen commented:
fixed style issues in the bench python script (as pointed out by the `flake8` check). Bus something is broken now when running it, looking into it now.

haukepetersen commented:
fixed, should be all good now.

PeterKietzmann commented:
nope, still not good:
`tests/periph_gpio/main.c:221: style (unsignedLessThanZero): Checking if unsigned variable 'runs' is less than zero.`

haukepetersen commented:
ups, missed that one.

haukepetersen commented:
next try (squashed fix right in)

kaspar030 commented:
Let's go!

kaspar030 commented:
I'll give travis another chance. Different cppcheck versions?

kaspar030 commented:
> I'll give travis another chance. Different cppcheck versions?

Yeah, and it complains about timex.h, thus unrelated.
